packages/mds-stream-processor/index.ts

```diff
@@ -15,40 +15,66 @@
  */
 
 import logger from '@mds-core/mds-logger'
+import { isError } from '@mds-core/mds-service-helpers'
 import { SingleOrArray } from '@mds-core/mds-types'
-import { asArray } from '@mds-core/mds-utils'
+import { asArray, ValidationError } from '@mds-core/mds-utils'
 import { StreamProcessorController, StreamSink, StreamSource, StreamTransform } from './@types'
 
 // StreamProcessor - Read from source, apply transform to each message, and write to sink
 export const StreamProcessor = <TMessageIn, TMessageOut>(
   source: StreamSource<TMessageIn>,
   transform: StreamTransform<TMessageIn, TMessageOut>,
-  sinks: SingleOrArray<StreamSink<TMessageOut>>
+  sinks: SingleOrArray<StreamSink<TMessageOut>>,
+  deadLetterSink?: StreamSink<TMessageIn>
 ): StreamProcessorController => {
-  const producers = asArray(sinks).map(sink => sink())
+  const sinkProducers = asArray(sinks).map(sink => sink())
+  const deadLetterProducer = deadLetterSink ? deadLetterSink() : null
+
   const consumer = source(async message => {
-    const transformed = await transform(message)
-    if (transformed) {
-      if (asArray(transformed).length > 0) {
-        try {
-          await Promise.all(producers.map(producer => producer.write(transformed)))
-        } catch (error) {
-          logger.error(`Error when writing to producer`, error)
-          await Promise.all(producers.map(producer => producer.shutdown()))
-          await Promise.all(producers.map(producer => producer.initialize()))
-          throw error
+    try {
+      const transformed = await transform(message)
+      if (transformed) {
+        if (asArray(transformed).length > 0) {
+          try {
+            await Promise.all(sinkProducers.map(producer => producer.write(transformed)))
+          } catch (error) {
+            logger.error(`Error when writing to producer`, error)
+            await Promise.all(sinkProducers.map(producer => producer.shutdown()))
+            await Promise.all(sinkProducers.map(producer => producer.initialize()))
+            throw error
+          }
+        }
+      }
+    } catch (error) {
+      logger.error(`Error when processing message`, { error, message })
+
+      if (isError(error, ValidationError)) {
```

Comment:
I would start with DLQ'ing messages with _all_ kinds of errors. You could take the same tactic I did, with `invalid_data` where there's a `{error_type, error_message, message}` written out.

Reply:
Easy enough to do!